format/midi/metaevents.go

```diff
@@ -0,0 +1,393 @@
+
+	case TypeInstrumentName:
+		metaevent("instrument_name", decodeInstrumentName)
+
+	case TypeLyric:
+		metaevent("lyric", decodeLyric)
+
+	case TypeMarker:
+		metaevent("marker", decodeMarker)
+
+	case TypeCuePoint:
+		metaevent("cue_point", decodeCuePoint)
+
+	case TypeProgramName:
+		metaevent("program_name", decodeProgramName)
+
+	case TypeDeviceName:
+		metaevent("device_name", decodeDeviceName)
+
+	case TypeMIDIChannelPrefix:
+		metaevent("midi_channel_prefix", decodeMIDIChannelPrefix)
+
+	case TypeMIDIPort:
+		metaevent("midi_port", decodeMIDIPort)
+
+	case TypeTempo:
+		metaevent("tempo", decodeTempo)
+
+	case TypeSMPTEOffset:
+		metaevent("smpte_offset", decodeSMPTEOffset)
+
+	case TypeTimeSignature:
+		metaevent("time_signature", decodeTimeSignature)
+
+	case TypeKeySignature:
+		metaevent("key_signature", decodeKeySignature)
+
+	case TypeEndOfTrack:
+		metaevent("end_of_track", decodeEndOfTrack)
+
+	case TypeSequencerSpecificEvent:
+		metaevent("sequencer_specific_event", decodeSequencerSpecificEvent)
+
+	default:
+		flush(d, "unknown meta event (%02x)", event)
+	}
+}
+
+func decodeSequenceNumber(d *decode.D) {
+	d.FieldUintFn("sequence_number", func(d *decode.D) uint64 {
+		seqno := uint64(0)
+
+		if data, err := vlf(d); err != nil {
+			d.Errorf("%v", err)
+		} else {
+			if len(data) > 0 {
+				seqno += uint64(data[0])
+			}
+
+			if len(data) > 1 {
+				seqno <<= 8
+				seqno += uint64(data[1])
+			}
+		}
+
+		return seqno
+	})
+}
+
+func decodeText(d *decode.D) {
+	d.FieldStrFn("text", vlstring)
+}
+
+func decodeCopyright(d *decode.D) {
+	d.FieldStrFn("copyright", vlstring)
+}
+
+func decodeTrackName(d *decode.D) {
+	d.FieldStrFn("name", vlstring)
+}
+
+func decodeInstrumentName(d *decode.D) {
+	d.FieldStrFn("instrument", vlstring)
+}
+
+func decodeLyric(d *decode.D) {
+	d.FieldStrFn("lyric", vlstring)
+}
+
+func decodeMarker(d *decode.D) {
+	d.FieldStrFn("marker", vlstring)
+}
+
+func decodeCuePoint(d *decode.D) {
+	d.FieldStrFn("cue", vlstring)
+}
+
+func decodeProgramName(d *decode.D) {
+	d.FieldStrFn("program", vlstring)
+}
+
+func decodeDeviceName(d *decode.D) {
+	d.FieldStrFn("device", vlstring)
+}
+
+func decodeMIDIChannelPrefix(d *decode.D) {
+	d.FieldUintFn("channel", func(d *decode.D) uint64 {
+		channel := uint64(0)
+
+		if data, err := vlf(d); err != nil {
+			d.Errorf("%v", err)
+		} else {
+			for _, b := range data {
+				channel <<= 8
+				channel |= uint64(b & 0x00ff)
+			}
+		}
+
+		return channel
+	})
+}
+
+func decodeMIDIPort(d *decode.D) {
+	d.FieldUintFn("port", func(d *decode.D) uint64 {
+		port := uint64(0)
+
+		if data, err := vlf(d); err != nil {
+			d.Errorf("%v", err)
+		} else {
+			for _, b := range data {
+				port <<= 8
+				port |= uint64(b & 0x00ff)
+			}
+		}
+
+		return port
+	})
+}
+
+func decodeTempo(d *decode.D) {
+	d.FieldUintFn("tempo", func(d *decode.D) uint64 {
+		tempo := uint64(0)
+
+		if data, err := vlf(d); err != nil {
+			d.Errorf("%v", err)
```

Comment:
Maybe some of these should be `d.Fatalf`? the difference is that `d.Errorf` can be ignored using `-o force=true`

Reply:
decodeMIDIPort, decodeMIDIPort and decodeSMPTEOffset looks quite similar, same kind of encoding? if so make sense to share somehow?

Reply:
Fixed in [e940f4766fe234cd25f32de03e1948ed184681df](https://github.com/wader/fq/pull/1004/commits/e940f4766fe234cd25f32de03e1948ed184681df)